universum/configuration_support.py

```diff
@@ -1,18 +1,120 @@
 import copy
 import os
 
+
 __all__ = [
+    "ProjectConfiguration",
     "Variations",
-    "combine",
     "set_project_root",
     "get_project_root"
 ]
 
-skip_attributes = {"children", "critical", "skip_numbering_level"}
+
+class ProjectConfiguration:
+    """
+    ProjectConfiguration is a collection of configurable project-specific data needed for Universum operations.
+    Legacy constructor supports dictionary data, but for static type checking please use new explicit constructor.
+    """
+
+    # pylint: disable-msg=too-many-locals
+    def __init__(self,
+                 name: str = '',
+                 command: Optional[List[str]] = None,
+                 environment: Optional[Dict[str, str]] = None,
+                 artifacts: str = '',
+                 report_artifacts: str = '',
+                 artifact_prebuild_clean: bool = False,
+                 directory: str = '',
+                 critical: bool = False,
+                 background: bool = False,
+                 finish_background: bool = False,
+                 code_report: bool = False,
+                 pass_tag: str = '',
+                 fail_tag: str = '',
+                 if_env_set: str = '',
+                 extras: Optional[Dict[str, str]] = None,
+                 **kwargs) -> None:  # explicit constructor for client's type safety
+        self.name: str = name
+        self.directory: str = directory
+        self.code_report: bool = code_report
+        self.command: List[str] = command if command else []
+        self.environment: Dict[str, str] = environment if environment else {}
+        self.artifacts: str = artifacts
+        self.report_artifacts: str = report_artifacts
+        self.artifact_prebuild_clean: bool = artifact_prebuild_clean
+        self.critical: bool = critical
+        self.background: bool = background
+        self.finish_background: bool = finish_background
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.skip_numbering_level: bool = False
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def replace_string(self, from_string: str, to_string: str) -> None:
+        self.command = [word.replace(from_string, to_string) for word in self.command]
+        self.artifacts = self.artifacts.replace(from_string, to_string)
+        self.report_artifacts = self.report_artifacts.replace(from_string, to_string)
+        self.directory = self.directory.replace(from_string, to_string)
+        # should we update extras? (user-defined)
+
+    def __repr__(self) -> str:
+        res = {k: v for k, v in self.__dict__.items() if v and k != 'extras'}
+        res.update(self.extras)
+        if len(self.command) == 1:  # command should be printed as one string, instead of list
```

Comment:
this way we don't have to use Union[str, List[str]] for command attribute, but can still print ProjectConfiguration cleanly